File: sdk/py/query.py
# pyre-strict
from dataclasses import dataclass
import typing
import serde_types as st
import bincode

class Algorithm:
    VARIANTS = []  # type: typing.Sequence[typing.Type[Algorithm]]

    def bincode_serialize(self) -> bytes:
        return bincode.serialize(self, Algorithm)

    @staticmethod
    def bincode_deserialize(input: bytes) -> 'Algorithm':
        v, buffer = bincode.deserialize(input, Algorithm)
        if buffer:
            raise st.DeserializationError("Some input bytes were not read");
        return v


@dataclass(frozen=True)
class Algorithm__EuclideanDistance(Algorithm):
    INDEX = 0  # type: int
    pass


@dataclass(frozen=True)
class Algorithm__DotProductSimilarity(Algorithm):
    INDEX = 1  # type: int
    pass


@dataclass(frozen=True)
class Algorithm__CosineSimilarity(Algorithm):
    INDEX = 2  # type: int
    pass

Algorithm.VARIANTS = [
    Algorithm__EuclideanDistance,
    Algorithm__DotProductSimilarity,
    Algorithm__CosineSimilarity,
]


@dataclass(frozen=True)
class Array:
    v: st.uint8
    dim: typing.Tuple[st.uint64]
    data: typing.Sequence[st.float32]

    def bincode_serialize(self) -> bytes:
        return bincode.serialize(self, Array)

    @staticmethod
    def bincode_deserialize(input: bytes) -> 'Array':
        v, buffer = bincode.deserialize(input, Array)
        if buffer:
            raise st.DeserializationError("Some input bytes were not read");
        return v


@dataclass(frozen=True)
class Predicate:
    key: str
    value: str
    op: "PredicateOp"

    def bincode_serialize(self) -> bytes:
        return bincode.serialize(self, Predicate)

    @staticmethod
    def bincode_deserialize(input: bytes) -> 'Predicate':
        v, buffer = bincode.deserialize(input, Predicate)
        if buffer:
            raise st.DeserializationError("Some input bytes were not read");
        return v


class PredicateCondition:
    VARIANTS = []  # type: typing.Sequence[typing.Type[PredicateCondition]]

    def bincode_serialize(self) -> bytes:
        return bincode.serialize(self, PredicateCondition)

    @staticmethod
    def bincode_deserialize(input: bytes) -> 'PredicateCondition':
        v, buffer = bincode.deserialize(input, PredicateCondition)
        if buffer:
            raise st.DeserializationError("Some input bytes were not read");
        return v


@dataclass(frozen=True)
class PredicateCondition__Value(PredicateCondition):
    INDEX = 0  # type: int
    value: "Predicate"


@dataclass(frozen=True)
class PredicateCondition__And(PredicateCondition):
    INDEX = 1  # type: int
    value: typing.Tuple["PredicateCondition", "PredicateCondition"]


@dataclass(frozen=True)
class PredicateCondition__Or(PredicateCondition):
    INDEX = 2  # type: int
    value: typing.Tuple["PredicateCondition", "PredicateCondition"]

PredicateCondition.VARIANTS = [
    PredicateCondition__Value,
    PredicateCondition__And,
    PredicateCondition__Or,
]


class PredicateOp:
    VARIANTS = []  # type: typing.Sequence[typing.Type[PredicateOp]]

    def bincode_serialize(self) -> bytes:
        return bincode.serialize(self, PredicateOp)

    @staticmethod
    def bincode_deserialize(input: bytes) -> 'PredicateOp':
        v, buffer = bincode.deserialize(input, PredicateOp)
        if buffer:
            raise st.DeserializationError("Some input bytes were not read");
        return v


@dataclass(frozen=True)
class PredicateOp__Equals(PredicateOp):
    INDEX = 0  # type: int
    pass


@dataclass(frozen=True)
class PredicateOp__NotEquals(PredicateOp):
    INDEX = 1  # type: int
    pass

PredicateOp.VARIANTS = [
    PredicateOp__Equals,
    PredicateOp__NotEquals,
]


class Query:
    VARIANTS = []  # type: typing.Sequence[typing.Type[Query]]

    def bincode_serialize(self) -> bytes:
        return bincode.serialize(self, Query)

    @staticmethod
    def bincode_deserialize(input: bytes) -> 'Query':
        v, buffer = bincode.deserialize(input, Query)
        if buffer:
            raise st.DeserializationError("Some input bytes were not read");
        return v


@dataclass(frozen=True)
class Query__CreateStore(Query):
    INDEX = 0  # type: int
    store: str
    dimension: st.uint64
    create_predicates: typing.Sequence[str]
    error_if_exists: bool


@dataclass(frozen=True)
class Query__GetKey(Query):
    INDEX = 1  # type: int
    store: str
    keys: typing.Sequence["Array"]


@dataclass(frozen=True)
class Query__GetPred(Query):
    INDEX = 2  # type: int
    store: str
    condition: "PredicateCondition"


@dataclass(frozen=True)
class Query__GetSimN(Query):
    INDEX = 3  # type: int
    store: str
    search_input: "Array"
    closest_n: st.uint64
    algorithm: "Algorithm"
    condition: typing.Optional["PredicateCondition"]


@dataclass(frozen=True)
class Query__CreateIndex(Query):
    INDEX = 4  # type: int
    store: str
    predicates: typing.Sequence[str]


@dataclass(frozen=True)
class Query__DropIndex(Query):
    INDEX = 5  # type: int
    store: str
    predicates: typing.Sequence[str]
    error_if_not_exists: bool


@dataclass(frozen=True)
class Query__Set(Query):
    INDEX = 6  # type: int
    store: str
    inputs: typing.Sequence[typing.Tuple["Array", typing.Dict[str, str]]]


@dataclass(frozen=True)
class Query__DelKey(Query):
    INDEX = 7  # type: int
    store: str
    keys: typing.Sequence["Array"]


@dataclass(frozen=True)
class Query__DelPred(Query):
    INDEX = 8  # type: int
    store: str
    condition: "PredicateCondition"


@dataclass(frozen=True)
class Query__DropStore(Query):
    INDEX = 9  # type: int
    store: str
    error_if_not_exists: bool


@dataclass(frozen=True)
class Query__InfoServer(Query):
    INDEX = 10  # type: int
    pass


@dataclass(frozen=True)
class Query__ListStores(Query):
    INDEX = 11  # type: int
    pass


@dataclass(frozen=True)
class Query__ListClients(Query):
    INDEX = 12  # type: int
    pass


@dataclass(frozen=True)
class Query__Ping(Query):
    INDEX = 13  # type: int
    pass

Query.VARIANTS = [
    Query__CreateStore,
    Query__GetKey,
    Query__GetPred,
    Query__GetSimN,
    Query__CreateIndex,
    Query__DropIndex,
    Query__Set,
    Query__DelKey,
    Query__DelPred,
    Query__DropStore,
    Query__InfoServer,
    Query__ListStores,
    Query__ListClients,
    Query__Ping,
]


@dataclass(frozen=True)
class ServerQuery:
    queries: typing.Sequence["Query"]

    def bincode_serialize(self) -> bytes:
        return bincode.serialize(self, ServerQuery)

    @staticmethod
    def bincode_deserialize(input: bytes) -> 'ServerQuery':
        v, buffer = bincode.deserialize(input, ServerQuery)
        if buffer:
            raise st.DeserializationError("Some input bytes were not read");
        return v

File: sdk/ahnlich-client-go/client/query.go
package client

import (
	"errors"
	"fmt"
	"sync"

	dbQuery "github.com/deven96/ahnlich/sdk/ahnlich-client-go/internal/db_query"
)

// TODO: Add Validation to the queries to avoid nil pointers

// AhnlichDBQueryBuilder builds queries based on input parameters
type AhnlichDBQueryBuilder struct {
	queries []dbQuery.Query
	mu      *sync.Mutex // To Ensure FIFO order of queries in the pipeline
}

// NewAhnlichAhnlichQueryBuilder creates a new instance of AhnlichDBQueryBuilder
func NewAhnlichDBQueryBuilder() *AhnlichDBQueryBuilder {
	return &AhnlichDBQueryBuilder{
		queries: make([]dbQuery.Query, 0),
		mu:      &sync.Mutex{},
	}
}

func (qb *AhnlichDBQueryBuilder) AddQuery(q dbQuery.Query) (err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("panic occurred when adding Query to query builder: %v", r) // Convert to AhnlichQueryBuilderException
		}
	}()
	qb.mu.Lock()
	defer qb.mu.Unlock()
	qb.queries = append(qb.queries, q)
	return nil
}

func (qb *AhnlichDBQueryBuilder) BuildCreateStoreQuery(storeName string, dimension uint64, predicates []string, nonLinearAlgorithm []dbQuery.NonLinearAlgorithm, errorIfExist bool) (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__CreateStore{
		Store:            storeName,
		Dimension:        dimension,
		CreatePredicates: predicates,
		NonLinearIndices: nonLinearAlgorithm,
		ErrorIfExists:    errorIfExist,
	})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildGetByKeysQuery(storeName string, keys []dbQuery.Array) (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__GetKey{
		Store: storeName,
		Keys:  keys,
	})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildGetByPredicateQuery(storeName string, condition dbQuery.PredicateCondition) (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__GetPred{
		Store:     storeName,
		Condition: condition,
	})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildGetBySimNQuery(storeName string, searchInput dbQuery.Array, closest_n uint64, algorithm dbQuery.Algorithm, condition *dbQuery.PredicateCondition) (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__GetSimN{
		Store:       storeName,
		SearchInput: searchInput,
		ClosestN:    closest_n,
		Algorithm:   algorithm,
		Condition:   condition,
	})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildDropPredicateIndexQuery(storeName string, predicates []string, errorIfNotExist bool) (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__DropPredIndex{
		Store:            storeName,
		Predicates:       predicates,
		ErrorIfNotExists: errorIfNotExist,
	})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildCreatePredicateIndexQuery(storeName string, predicates []string) (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__CreatePredIndex{
		Store:      storeName,
		Predicates: predicates,
	})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildSetQuery(storeName string, inputs []struct {
	Field0 dbQuery.Array
	Field1 map[string]dbQuery.MetadataValue
}) (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__Set{
		Store:  storeName,
		Inputs: inputs,
	})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildDeleteKeysQuery(storeName string, keys []dbQuery.Array) (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__DelKey{
		Store: storeName,
		Keys:  keys,
	})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildDeletePredicateQuery(storeName string, condition dbQuery.PredicateCondition) (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__DelPred{
		Store:     storeName,
		Condition: condition,
	})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildDropStoreQuery(storeName string, errorIfNotExist bool) (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__DropStore{
		Store:            storeName,
		ErrorIfNotExists: errorIfNotExist,
	})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildListStoresQuery() (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__ListStores{})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildInfoServerQuery() (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__InfoServer{})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildListClientsQuery() (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__ListClients{})
	return qb, err
}

func (qb *AhnlichDBQueryBuilder) BuildPingQuery() (*AhnlichDBQueryBuilder, error) {
	err := qb.AddQuery(&dbQuery.Query__Ping{})
	return qb, err
}

// Reset drops all the queries in the query builder
func (qb *AhnlichDBQueryBuilder) Reset() error {
	qb.queries = make([]dbQuery.Query, 0)
	return nil
}

// ParseBuildQueryToServer parses the queries and builds a server query and drops the queries from the query builder
func (qb *AhnlichDBQueryBuilder) ParseBuildQueryToServerQuery() (*dbQuery.ServerQuery, error) {
	if len(qb.queries) == 0 {
		return nil, errors.New("must have atleast one request to be processed")
	}
	queries := make([]dbQuery.Query, len(qb.queries))
	copy(queries, qb.queries)
	qb.Reset()
	return &dbQuery.ServerQuery{Queries: queries}, nil
}

// Create a DB query array type from a slice of float32
func MakeDBQueryArrayType(data []float32, v uint8) dbQuery.Array {
	data32 := make([]float32, len(data))
	for i, d := range data {
		data32[i] = float32(d)
	}
	dimensions := struct{ Field0 uint64 }{Field0: uint64(len(data))}
	return dbQuery.Array{
		V:    v,
		Dim:  dimensions,
		Data: data32,
	}
}

// Create a DB query map of metadata type from a map of string
func MakeDBQueryMetaDataType(data map[string]string) map[string]dbQuery.MetadataValue {
	metadata := make(map[string]dbQuery.MetadataValue)
	for k, v := range data {
		val := dbQuery.MetadataValue__RawString(v)
		metadata[k] = &val
	}
	return metadata
}

// Create a DB query map of metadata type from a map of binary data in the form of a slice of uint8
func MakeDBQueryMetaDataTypeBinary(data map[string][]uint8) map[string]dbQuery.MetadataValue {
	metadata := make(map[string]dbQuery.MetadataValue)
	for k, v := range data {
		val := dbQuery.MetadataValue__Binary(v)
		metadata[k] = &val
	}
	return metadata
}
